File: ui/closures/net.bluemind.deferredaction.closure/src/main/javascript/net/bluemind/deferredaction/service/DeferredActionService.js
/**
 * BEGIN LICENSE
 * Copyright © Blue Mind SAS, 2012-2019
 *
 * This file is part of BlueMind. BlueMind is a messaging and collaborative
 * solution.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of either the GNU Affero General Public License as
 * published by the Free Software Foundation (version 3 of the License).
 *
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 *
 * See LICENSE.txt
 * END LICENSE
 *
 * @format
 */

goog.provide("net.bluemind.deferredaction.service.DeferredActionService");

goog.require("net.bluemind.container.service.ContainerService");
goog.require("net.bluemind.deferredaction.api.DeferredActionClient");
goog.require("net.bluemind.mvp.helper.ServiceHelper");
goog.require("goog.array");

var CACHE_EXPIRATION = 9 * 60 * 1000;
/**
 * @constructor
 * @extends {goog.events.EventTarget}
 */
net.bluemind.deferredaction.service.DeferredActionService = function(ctx) {
    goog.base(this);
    this.ctx = ctx;
    this.cache_ = { ts: 0, values: [] };
    this.cs_ = new net.bluemind.container.service.ContainerService(ctx, "deferredaction");
    this.css_ = new net.bluemind.container.service.ContainersService(ctx, "deferredaction");
};

goog.inherits(net.bluemind.deferredaction.service.DeferredActionService, goog.events.EventTarget);

net.bluemind.deferredaction.service.DeferredActionService.prototype.isLocal = function() {
    return this.cs_.available();
};

/**
 * Execute the right method depending on application state.
 *
 * @param {Object.<string, Function>} states
 * @param {Array.<*>} params Array of function parameters
 * @return {!goog.Promise<*>}
 */
net.bluemind.deferredaction.service.DeferredActionService.prototype.handleByState = function(states, params) {
    var localState = [];
    if (this.cs_.available()) {
        localState.push("local");
    }
    if (this.ctx.online) {
        localState.push("remote");
    }
    try {
        return net.bluemind.mvp.helper.ServiceHelper.handleByState(this.ctx, this, states, params, localState);
    } catch (e) {
        return net.bluemind.mvp.helper.ServiceHelper.handleByState(this.ctx, this, states, params, ["remote"]);
    }
};

net.bluemind.deferredaction.service.DeferredActionService.prototype.getItemsByDate = function(executionDate) {
    return this.handleByState(
        {
            "local": this.getItemsByDateLocal,
            "remote": this.getItemsByDateRemote
        },
        [executionDate]
    );
};

net.bluemind.deferredaction.service.DeferredActionService.prototype.getItemsByDateRemote = function(executionDate) {
    var promise;
    if (this.cache_.ts < executionDate) {
        var container = "deferredaction-" + this.ctx.user["uid"];
        var client = new net.bluemind.deferredaction.api.DeferredActionClient(this.ctx.rpc, "", container);
        promise = client.getByActionId("EVENT", executionDate + CACHE_EXPIRATION).then(
            function(items) {
                this.cache_.ts = executionDate + CACHE_EXPIRATION;
                this.cache_.values = items;
                return items;
            },
            null,
            this
        );
    } else {
        promise = goog.Promise.resolve(this.cache_.values);
    }
    return promise.then(function(items) {
        return goog.array.filter(items, function(item) {
            return item["value"]["executionDate"] < executionDate;
        });
    });
};

net.bluemind.deferredaction.service.DeferredActionService.prototype.getItemsByDateLocal = function(executionDate) {
    var query = [["value.executionDate", "<", executionDate]];
    return this.css_.searchItems(query).then(function(items) {
        return items.filter(function(item) {
            return item && item["value"]["actionId"] === "EVENT";
        });
    });
};

net.bluemind.deferredaction.service.DeferredActionService.prototype.getItem = function(id) {
    return this.handleByState(
        {
            "local": this.getItemLocal,
            "remote": function() {
                return goog.Promise.resolve(null);
            }
        },
        [id]
    );
};

net.bluemind.deferredaction.service.DeferredActionService.prototype.getItemLocal = function(id) {
    var containerId = "deferredaction-" + this.ctx.user["uid"];
    return this.cs_.getItem(containerId, id);
};

net.bluemind.deferredaction.service.DeferredActionService.prototype.deleteItem = function(item) {
    return this.handleByState(
        {
            "local": this.deleteItemLocal,
            "remote": this.deleteItemRemote
        },
        [item]
    );
};

net.bluemind.deferredaction.service.DeferredActionService.prototype.deleteItemLocal = function(item) {
    var containerId = "deferredaction-" + this.ctx.user["uid"];
    return this.cs_.deleteItem(containerId, item["uid"]);
};

net.bluemind.deferredaction.service.DeferredActionService.prototype.deleteItemRemote = function(item) {
    var container = "deferredaction-" + this.ctx.user["uid"];
    var client = new net.bluemind.deferredaction.api.DeferredActionClient(this.ctx.rpc, "", container);
    goog.array.removeIf(this.cache_.values, function(value) {
        return value["uid"] === item["uid"];
    });
    return client.delete_(item["uid"]);
};

net.bluemind.deferredaction.service.DeferredActionService.prototype.createItem = function(item) {
    return this.handleByState(
        {
            "local": this.createItemLocal,
            "remote": this.createItemRemote
        },
        [item]
    );
};

net.bluemind.deferredaction.service.DeferredActionService.prototype.createItemLocal = function(item) {
    item["container"] = "deferredaction-" + this.ctx.user["uid"];
    return this.cs_.storeItem(item);
};

net.bluemind.deferredaction.service.DeferredActionService.prototype.createItemRemote = function(item) {
    var container = "deferredaction-" + this.ctx.user["uid"];
    var client = new net.bluemind.deferredaction.api.DeferredActionClient(this.ctx.rpc, "", container);
    return client.create(item["uid"], item["value"]).then(
        function() {},
        function() {}
    );
};

File: parent/deferredaction/net.bluemind.deferredaction.service/src/net/bluemind/deferredaction/service/internal/DeferredActionService.java
/* BEGIN LICENSE
 * Copyright © Blue Mind SAS, 2012-2019
 *
 * This file is part of BlueMind. BlueMind is a messaging and collaborative
 * solution.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of either the GNU Affero General Public License as
 * published by the Free Software Foundation (version 3 of the License).
 *
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 *
 * See LICENSE.txt
 * END LICENSE
 */
package net.bluemind.deferredaction.service.internal;

import java.sql.SQLException;
import java.util.Date;
import java.util.List;
import java.util.UUID;

import javax.sql.DataSource;

import net.bluemind.core.api.fault.ServerFault;
import net.bluemind.core.container.model.Container;
import net.bluemind.core.container.model.ContainerChangelog;
import net.bluemind.core.container.model.ContainerChangeset;
import net.bluemind.core.container.model.ItemChangelog;
import net.bluemind.core.container.model.ItemFlagFilter;
import net.bluemind.core.container.model.ItemValue;
import net.bluemind.core.container.model.ItemVersion;
import net.bluemind.core.container.model.acl.Verb;
import net.bluemind.core.container.persistence.ContainerStore;
import net.bluemind.core.container.service.ChangeLogUtil;
import net.bluemind.core.container.service.internal.RBACManager;
import net.bluemind.core.rest.BmContext;
import net.bluemind.core.sanitizer.Sanitizer;
import net.bluemind.core.validator.Validator;
import net.bluemind.deferredaction.api.DeferredAction;
import net.bluemind.deferredaction.api.IInternalDeferredAction;
import net.bluemind.deferredaction.persistence.DeferredActionStore;

public class DeferredActionService implements IInternalDeferredAction {
	private final Container container;
	private final BmContext context;
	private final ContainerDeferredActionStoreService storeService;
	private final DeferredActionStore deferredActionStore;
	private final Validator validator;
	private final Sanitizer sanitizer;

	public DeferredActionService(Container container, DataSource dataSource, BmContext context) {
		this.container = container;
		this.context = context;
		this.deferredActionStore = new DeferredActionStore(dataSource, container);
		this.storeService = new ContainerDeferredActionStoreService(dataSource, context.getSecurityContext(), container,
				deferredActionStore);
		this.validator = new Validator(context);
		this.sanitizer = new Sanitizer(context);
	}

	@Override
	public void create(DeferredAction deferredAction) throws ServerFault {
		byte[] u = (deferredAction.reference + deferredAction.executionDate.toString()).getBytes();
		String uid = UUID.nameUUIDFromBytes(u).toString();
		create(uid, deferredAction);
	}

	@Override
	public void create(String uid, DeferredAction deferredAction) throws ServerFault {
		ItemValue<DeferredAction> deferredActionItem = ItemValue.create(uid, deferredAction);
		deferredActionItem.displayName = displayName(deferredAction);
		create(deferredActionItem);
	}

	private void create(ItemValue<DeferredAction> deferredActionItem) throws ServerFault {
		RBACManager.forContext(context).forContainer(container).check(Verb.Write.name());
		DeferredAction deferredAction = deferredActionItem.value;
		sanitizer.create(deferredAction);
		validator.create(deferredAction);

		storeService.create(deferredActionItem.item(), deferredAction);
	}

	@Override
	public void update(String uid, DeferredAction deferredAction) throws ServerFault {
		ItemValue<DeferredAction> deferredActionItem = ItemValue.create(uid, deferredAction);
		deferredActionItem.displayName = displayName(deferredAction);
		update(deferredActionItem);
	}

	private void update(ItemValue<DeferredAction> deferredActionItem) throws ServerFault {
		RBACManager.forContext(context).forContainer(container).check(Verb.Write.name());
		String uid = deferredActionItem.uid;
		DeferredAction deferredAction = deferredActionItem.value;
		ItemValue<DeferredAction> current = storeService.get(uid, null);
		sanitizer.update(current, deferredAction);
		validator.update(current, deferredAction);

		storeService.update(deferredActionItem.item(), deferredActionItem.displayName, deferredAction);
	}

	@Override
	public void delete(String uid) throws ServerFault {
		RBACManager.forContext(context).forContainer(container).check(Verb.Write.name());

		storeService.delete(uid);
	}

	@Override
	public void deleteAll() throws ServerFault {
		RBACManager.forContext(context).forContainer(container).check(Verb.Write.name());

		storeService.deleteAll();
	}

	@Override
	public DeferredAction get(String uid) throws ServerFault {
		ItemValue<DeferredAction> item = getComplete(uid);
		return (item != null) ? item.value : null;
	}

	@Override
	public ItemValue<DeferredAction> getComplete(String uid) throws ServerFault {
		RBACManager.forContext(context).forContainer(container).check(Verb.Read.name());

		return storeService.get(uid, null);
	}

	@Override
	public List<ItemValue<DeferredAction>> getByActionId(String actionId, Long to) throws ServerFault {
		RBACManager.forContext(context).forContainer(container).check(Verb.Read.name());

		return storeService.getByActionId(actionId, new Date(to));
	}

	@Override
	public List<ItemValue<DeferredAction>> getByReference(String reference) throws ServerFault {
		RBACManager.forContext(context).forContainer(container).check(Verb.Read.name());

		return storeService.getByReference(reference);
	}

	@Override
	public ItemChangelog itemChangelog(String itemUid, Long since) throws ServerFault {
		RBACManager.forContext(context).forContainer(container).check(Verb.Read.name());

		return ChangeLogUtil.getItemChangeLog(itemUid, since, context, storeService, container.domainUid);
	}

	@Override
	public ContainerChangelog containerChangelog(Long since) throws ServerFault {
		RBACManager.forContext(context).forContainer(container).check(Verb.Read.name());

		return storeService.changelog(since, Long.MAX_VALUE);
	}

	@Override
	public ContainerChangeset<String> changeset(Long since) throws ServerFault {
		RBACManager.forContext(context).forContainer(container).check(Verb.Read.name());

		return storeService.changeset(since, Long.MAX_VALUE);
	}

	@Override
	public ContainerChangeset<Long> changesetById(Long since) throws ServerFault {
		RBACManager.forContext(context).forContainer(container).check(Verb.Read.name());

		return storeService.changesetById(since, Long.MAX_VALUE);
	}

	@Override
	public ContainerChangeset<ItemVersion> filteredChangesetById(Long since, ItemFlagFilter filter) throws ServerFault {
		RBACManager.forContext(context).forContainer(container).check(Verb.Read.name());

		return storeService.changesetById(since, filter);
	}

	@Override
	public long getVersion() throws ServerFault {
		RBACManager.forContext(context).forContainer(container).check(Verb.Read.name());

		return storeService.getVersion();
	}

	private String displayName(DeferredAction deferredAction) {
		return deferredAction.actionId + "-" + deferredAction.executionDate.toString();
	}

	@Override
	public List<ItemValue<DeferredAction>> multipleGet(List<String> uids) throws ServerFault {
		RBACManager.forContext(context).forContainer(container).check(Verb.Read.name());

		return storeService.getMultiple(uids);
	}

	@Override
	public void xfer(String serverUid) throws ServerFault {
		DataSource ds = context.getMailboxDataSource(serverUid);
		ContainerStore cs = new ContainerStore(null, ds, context.getSecurityContext());
		Container c;
		try {
			c = cs.get(container.uid);
		} catch (SQLException e) {
			throw ServerFault.sqlFault(e);
		}
		storeService.xfer(ds, c, new DeferredActionStore(ds, c));
	}

	@Override
	public void restore(ItemValue<DeferredAction> deferredActionItem, boolean isCreate) {
		if (isCreate) {
			create(deferredActionItem);
		} else {
			update(deferredActionItem);
		}
	}

}
